In [45]:
from ISLP import load_data
import statsmodels.formula.api as smf

In [46]:
Boston = load_data("Carseats")

### a) Fit a multiple regression model to predict Sales using Price, Urban, and US.

In [47]:
X = Boston[["Price", "Urban", "US", "Sales"]]
X.head()

Price Urban   US  Sales
0    120   Yes  Yes   9.50
1     83   Yes  Yes  11.22
2     80   Yes  Yes  10.06
3     97   Yes  Yes   7.40
4    128   Yes   No   4.15

In [48]:
model = smf.ols(formula="Sales ~ Price + Urban + US", data=X)
result = model.fit()

In [49]:
print(result.summary())

                            OLS Regression Results                            
Dep. Variable:                  Sales   R-squared:                       0.239
Model:                            OLS   Adj. R-squared:                  0.234
Method:                 Least Squares   F-statistic:                     41.52
Date:                Tue, 31 Oct 2023   Prob (F-statistic):           2.39e-23
Time:                        20:35:56   Log-Likelihood:                -927.66
No. Observations:                 400   AIC:                             1863.
Df Residuals:                     396   BIC:                             1879.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept       13.0435      0.651     20.036   

### (b) Provide an interpretation of each coefficient in the model. Be careful—some of the variables in the model are qualitative!

* El coeficiente de "Price" es -0.0545, es decir, por cada unidad que aumenta el precio, las ventas disminuyen en 0.0545 unidades.
* Debido a que usamos variables categóricas para las columnas "Urban" y "US", para cada columna, el modelo utiliza "k-1" niveles. Como la columna "Urban" tiene 2 niveles ("Yes" y "No"), el modelo solo utiliza 1 nivel para el valor "Yes". Lo mismo sucede para la columna "US". Es decir que para ambos casos, el modelo utiliza  el nivel "No" como referencia.
  * El coeficiente de "Urban[T.Yes]" significa que "un cambio desde 'No' hasta 'Yes' disminuye en 0.0219 unidades el valor de la venta".
  * El coeficiente de "US[T.Yes]" significa que "un cambio desde 'No' hasta 'Yes' incrementa en 1.2006 unidades el valor de la venta".

### (c) Write out the model in equation form, being careful to handle the qualitative variables properly.

Sales = 13.0435 - 0.0545 * Price - 0.0219 * Urban[T.Yes] + 1.2006 * US[T.Yes]

### (d) For which of the predictors can you reject the null hypothesis H0 : Bj = 0?

El valor p para todos los predictores excepto para "Urban[T.Yes]" es muy pequeño, por lo que podemos rechazar la hipótesis nula para todos los predictores excepto para "Urban[T.Yes]".

### (e) On the basis of your response to the previous question, fit a smaller model that only uses the predictors for which there is evidence of association with the outcome.

In [51]:
model2 = smf.ols(formula="Sales ~ Price + US", data=X)
result2 = model2.fit()
print(result2.summary())

                            OLS Regression Results                            
Dep. Variable:                  Sales   R-squared:                       0.239
Model:                            OLS   Adj. R-squared:                  0.235
Method:                 Least Squares   F-statistic:                     62.43
Date:                Tue, 31 Oct 2023   Prob (F-statistic):           2.66e-24
Time:                        20:36:04   Log-Likelihood:                -927.66
No. Observations:                 400   AIC:                             1861.
Df Residuals:                     397   BIC:                             1873.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     13.0308      0.631     20.652      0.0

### (f) How well do the models in (a) and (e) fit the data?

Ambos modelos dan un R^2 de 0.239, por lo que ambos modelos explican el 23.9% de la variabilidad de los datos. Sin embargo, por el principio de parsimonia, nos podemos quedar con el segundo modelo.

### (g) Using the model from (e), obtain 95 % confidence intervals for the coefficient(s).

* Para el intercepto, el intervalo de confianza del 95% es [11.790, 14.271]
* Para "US[T.Yes]" el intervalo de confianza del 95% es [0.692, 1.708]
* Para "Price" el intervalo de confianza del 95% es [-0.065, -0.044]

### (h) Is there evidence of outliers or high leverage observations in the model from (e)?

Debido a que la kurtosis es casi 3 (su valor real es 2.895), entonces no hay evidencia de outliers.